### DISCUSSION TITLE
Consistent reaction behavior between Python and Java

### DISCUSSION THREAD

**Discussion by MartinHoffmannJena** [rdkit/rdkit#8067] [Q&A] [answered]
Hi,

I'm trying to figure out how to get consistent reaction behavior between the Python and JavaWrapper versions of Rdkit.

I have the following reactionSMARTs: "[#6:1]-[#8:2]>>[#6:1]-[#7:2]", which just depicts CO -> CN.

In Python I can do

```
mol1 =Chem.MolFromSmiles("CCCCCCO")

rxn=rdChemReactions.ReactionFromSmarts("[#6:1]-[#8:2]>>[#6:1]-[#7:2]")

print(Chem.MolToSmiles(rxn.RunReactant((mol1),0)[0][0]))
```

And then my output is `CCCCCCN`, which is what I'm expecting.

Now in Java:

```
ChemicalReaction r = ChemicalReaction.ReactionFromSmarts("[#6:1]-[#8:2]>>[#6:1]-[#7:2]");

RWMol test = RWMol.MolFromSmiles("CCCCCCO");

r.runReactant(test,false);

System.out.println(r.getProducts().get(0).MolToSmiles());
```

My output here however is `[CH3:1][NH2:2]`, the unmatched Atoms of my reactant were lost, even though I set `boolean removeUnmatchedAtoms` to `false` in `runReactant()`. How can I change this behaviour?

Thank you very much!

**Comment by bp-kelley** [comment, marked as answer]
getProduct returns the product template from the reaction, not the products of the reaction.

the products, as with python, are returned by runReactant(s)

**Reply by MartinHoffmannJena** [reply]
![image](https://github.com/user-attachments/assets/732573aa-c31e-44f1-a0fd-cd187a8d6ccf)

Thank you for the response, runReactant has boolean as the return value which is why I assumed the reaction product is the actual product.

As it turns out, runReactant (without the s) is in place automatically in Java, so the result was in `test.MolToSmiles()`

Thank you!

**Comment by bp-kelley** [comment]
Example from

https://github.com/rdkit/rdkit/blob/master/Code/JavaWrappers/gmwrapper/src-test/org/RDKit/ChemReactionTests.java


```
rxn.initReactantMatchers();
ROMol_Vect_Vect prods = rxn.runReactants(reacts);
assertEquals( 1,prods.size() );
assertEquals( 1,prods.get(0).size() );
assertEquals( 3,prods.get(0).get(0).getNumAtoms() ); 
```
